charmcraft/config.py

```diff
@@ -296,24 +296,34 @@ class AnalysisConfig(ModelConfigDefaults, allow_population_by_field_name=True):
 class Config(ModelConfigDefaults, validate_all=False):
     """Definition of charmcraft.yaml configuration."""
 
+    # this needs to go before 'parts', as it used by the validator
+    project: Project
+
     type: str
     charmhub: CharmhubConfig = CharmhubConfig()
     parts: Optional[Dict[str, Any]]
     bases: Optional[List[BasesConfiguration]]
     analysis: AnalysisConfig = AnalysisConfig()
 
-    project: Project
-
     @pydantic.validator("type")
     def validate_charm_type(cls, charm_type):
         """Verify charm type is valid with exception when instantiated without YAML."""
         if charm_type not in ["bundle", "charm"]:
             raise ValueError("must be either 'charm' or 'bundle'")
         return charm_type
 
-    @pydantic.validator("parts", pre=True)
-    def validate_special_parts(cls, parts):
-        """Verify parts type (craft-parts will re-validate this)."""
+    @pydantic.validator("parts", pre=True, always=True)
+    def validate_special_parts(cls, parts, values):
+        """Verify parts type (craft-parts will re-validate the schemas for the plugins)."""
+        if "type" not in values:
+            # we need 'type' to be set in this validator; if not there it's an error in
+            # the schema anyway, so the whole loading will fail
+            return
```

Comment:
That `type` is present and ok is validated in the method just above, so if this expression is `True` the schema validation will fail and a proper error will be shown to the user, no need to add another error here.

Note this is different for the case of Charm Plugin Properties where for `charm-requirements` it validates that there is ` source`, because that class is not really validating that `source` is there, that is responsibility of somebody else (the plugin mechanism in craft-parts).

Reply:
I just improved the comment in code.